packages/api/endpoints/launchpadSaml.js

```diff
@@ -136,20 +141,65 @@ const auth = async (req, res) => {
         const Location = `${req.body.RelayState}/?token=${LaunchpadJwtToken}`;
         return res.redirect(Location);
       })
-      .catch((error) => res.boom.badRequest('Could not build JWT from SAML response', error));
+      .catch((error) =>
+        res.boom.badRequest('Could not build JWT from SAML response', error));
   });
 };
 
-const notImplemented = async (req, res) => res.boom.notImplemented(
-  `endpoint: "${req.path}" not implemented. Login with launchpad.`
-);
+/**
+ * Helper to pull the url the client sent from a request that has been updated
+ * with middleware to include the event context.
+ * @param {Object} req - express request object
+ * @returns {Object} - The url the client visited to generate the request.
+ */
+const urlFromRequest = (req) =>
+  `${req.protocol}://${req.get('host')}${req.apiGateway.event.requestContext.path}`;
+
+/**
+ * helper to grab stageName
+ *
+ * @param {Object} req - express request object
+ * @returns {string} - stage name of apigateway
+ */
+const stageNameFromRequest = (req) => req.apiGateway.event.requestContext.stage;
+
+/**
+ * SAML Token endpoint.
+ *
+ * Simply returns the token received as a query parameter or redirects to saml
+ * login to authenticate.
+ * @param {Object} req - express request
+ * @param {Object} res - express response
+ * @returns {Object} - Either JWToken presented as a query string in the
+ * request or a redirect back to saml/login endpoing to receive the token.
+ */
+const samlToken = async (req, res) => {
+  let relayState;
+  let stageName;
+  try {
+    relayState = encodeURIComponent(urlFromRequest(req));
+    stageName = stageNameFromRequest(req);
+    if (!relayState || !stageName) throw new Error('bad');
```

Comment:
We should probably throw a better error here? And then log it in the catch below?

Reply:
My thought was that this was just a catch to make sure that the express server was using the middleware so that the full path to this endpoint (how token was called) and the stagename were available.  

I wouldn't expect either of these to ever not succeed except if you're writing a test and don't explicitly include the x-apigateway-context and x-apigateway-event headers.  If the app is deployed with the middleware against AWS's apiGateway I think we should always have these values.

This 'bad' error is caught immediately and we return an expecationFailed error.    I suppose we could add information onto that message if you thought it would be useful.  
urlFromRequest will throw it's own error if the app is incorrectly configured (a TypeError on Event because apiGateway will be undefined.)

Reply:
I guess I just reacted to seeing a non-specific error being thrown. Ultimately, it seems like someone later could refactor the code and to include a re-throw and the error here might surface. So if nothing else I would advocate just changing the error message. But maybe it's YAGNI. Totally your call

Reply:
ce8709c12